pkg/operator/quorumguardcontroller/quorumguardcontroller.go

```diff
@@ -0,0 +1,248 @@
+package quorumguardcontroller
+
+import (
+	"context"
+	"fmt"
+	configv1listers "github.com/openshift/client-go/config/listers/config/v1"
+	"gopkg.in/yaml.v2"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"strconv"
+	"time"
+
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/etcd_assets"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/ceohelpers"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/operatorclient"
+)
+
+const (
+	EtcdGuardDeploymentName            = "etcd-quorum-guard"
+	infrastructureClusterName          = "cluster"
+	clusterConfigName                  = "cluster-config-v1"
+	clusterConfigKey                   = "install-config"
+	clusterConfigNamespace             = "kube-system"
+	pdbDeploymentMaxUnavailableDefault = 1
+)
+
+type replicaCountDecoder struct {
+	ControlPlane struct {
+		Replicas string `yaml:"replicas,omitempty"`
+	} `yaml:"controlPlane,omitempty"`
+}
+
+// QuorumGuardController watches the etcd quorum guard deployment, create if not exists
+type QuorumGuardController struct {
+	operatorClient       v1helpers.OperatorClient
+	kubeClient           kubernetes.Interface
+	podLister            corev1listers.PodLister
+	nodeLister           corev1listers.NodeLister
+	infrastructureLister configv1listers.InfrastructureLister
+	clusterTopology      configv1.TopologyMode
+	replicaCount         int
+}
+
+func NewQuorumGuardController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	kubeInformers v1helpers.KubeInformersForNamespaces,
+	eventRecorder events.Recorder,
+	infrastructureLister configv1listers.InfrastructureLister,
+) factory.Controller {
+	c := &QuorumGuardController{
+		operatorClient:       operatorClient,
+		kubeClient:           kubeClient,
+		podLister:            kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Lister(),
+		nodeLister:           kubeInformers.InformersFor("").Core().V1().Nodes().Lister(),
+		infrastructureLister: infrastructureLister,
+		replicaCount:         0,
+	}
+	return factory.New().ResyncEvery(time.Minute).WithInformers(
```

Comment:
I think this probably a little aggressive CVO uses 10 minutes I think we can as well.